File: 05-hashmap/03-word-pattern/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if (pattern.length() != words.length) return false;

        Map<Character, String> mapPW = new HashMap<>();
        Map<String, Character> mapWP = new HashMap<>();
        for (int i = 0; i < pattern.length(); i++) {
            char ch = pattern.charAt(i);
            String word = words[i];
            if (mapPW.containsKey(ch) && !mapPW.get(ch).equals(word)) return false;
            if (mapWP.containsKey(word) && mapWP.get(word) != ch) return false;
            mapPW.put(ch, word);
            mapWP.put(word, ch);
        }
        return true;
    }
}

File: 05-hashmap/03-word-pattern/Solution.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False
        map_pw: dict[str, str] = {}
        map_wp: dict[str, str] = {}
        for ch, word in zip(pattern, words):
            if ch in map_pw and map_pw[ch] != word:
                return False
            if word in map_wp and map_wp[word] != ch:
                return False
            map_pw[ch] = word
            map_wp[word] = ch
        return True
